### PR TITLE
Add new subsample function to occupancy grid.

### PR DESCRIPTION
**Pull request by schneith**

### CODE REVIEW

File: aslam_cv_common/include/aslam/common/occupancy-grid-inl.h [hunk 1]
```diff
@@ -224,19 +224,65 @@ WeightedOccupancyGrid<PointType>::setConstantWeightForAllPointsInGrid(
...
     CHECK_EQ(current_num_points_, num_pts);
- // #endif
+#endif
 
   return current_num_points_;
 }
 
+template<typename PointType>
+typename WeightedOccupancyGrid<PointType>::GridCoordinates
+WeightedOccupancyGrid<PointType>::getFullestGridCell() const {
```
Comment (dymczykm) [inline aslam_cv_common/include/aslam/common/occupancy-grid-inl.h]: MostOccupied?

File: aslam_cv_common/include/aslam/common/occupancy-grid-inl.h [hunk 2]
```diff
@@ -224,19 +224,65 @@ WeightedOccupancyGrid<PointType>::setConstantWeightForAllPointsInGrid(
...
+        max_size = grid_[r][c].size();
+        coords = GridCoordinates(r, c);
+      }
+    }
+  }
+  return coords;
+}
+
+template<typename PointType>
+void WeightedOccupancyGrid<PointType>::removePointsFromLargestCellsUntilSize(
```
Comment (dymczykm) [inline aslam_cv_common/include/aslam/common/occupancy-grid-inl.h]: FromMostOccupiedCells...?

File: aslam_cv_common/include/aslam/common/occupancy-grid-inl.h [hunk 3]
```diff
@@ -224,19 +224,65 @@ WeightedOccupancyGrid<PointType>::setConstantWeightForAllPointsInGrid(
...
+    size_t max_total_num_points) {
+  CHECK_GT(max_total_num_points, 0u);
+  while (getNumPoints() > max_total_num_points) {
+    const GridCoordinates fullest_cell_coords = getFullestGridCell();
```
Comment (dymczykm) [inline aslam_cv_common/include/aslam/common/occupancy-grid-inl.h]: most_occupied_coords

File: aslam_cv_common/include/aslam/common/occupancy-grid-inl.h [hunk 4]
```diff
@@ -224,19 +224,65 @@ WeightedOccupancyGrid<PointType>::setConstantWeightForAllPointsInGrid(
...
+    const size_t max_cell_size = getGridCell(fullest_cell_coords).size();
```
Comment (dymczykm) [inline aslam_cv_common/include/aslam/common/occupancy-grid-inl.h]: max_cell_count

File: aslam_cv_common/include/aslam/common/occupancy-grid.h [hunk 5]
```diff
@@ -115,6 +116,13 @@ class WeightedOccupancyGrid {
   /// Remove points from cells that contain more points than specified. Points with the lowest
   /// score will be removed first.
   size_t removeWeightedPointsFromOverfullCells(size_t max_points_per_cell);
+
+  size_t removeWeightedPointsFromOverfullCell(
+      const GridCoordinates& grid_coords, size_t max_points_per_cell);
+
+  /// Remove the weakest point from the largest cell until the total number of
```
Comment (dymczykm) [inline aslam_cv_common/include/aslam/common/occupancy-grid.h]: from the most occupied cell

File: aslam_cv_common/include/aslam/common/occupancy-grid-inl.h [hunk 6]
```diff
@@ -224,19 +224,65 @@ WeightedOccupancyGrid<PointType>::setConstantWeightForAllPointsInGrid(
...
+  CHECK_GT(max_points_per_cell, 0u);
+
+  PointList& cell = getGridCell(grid_coords);
+  if (cell.size() <= max_points_per_cell) {
+    return 0u;
+  }
+
+  // Remove the points with the lowest score.
+  size_t num_removed = cell.size() - max_points_per_cell;
+  std::sort(cell.begin(), cell.end(), std::greater<PointType>());
```
Comment (schneith) [inline aslam_cv_common/include/aslam/common/occupancy-grid-inl.h]: partial_sort

**Pull request merged**